File: mlp.py
import numpy as np
import time
from PIL import Image, ImageDraw

class MLP:
    def __init__(self):
        self.hidden_size = 4
        
        # C++의 mt19937과 동일한 시드값 사용
        np.random.seed(5489)  # mt19937의 기본 시드값
        
        # float64(C++의 double과 동일) 타입 명시
        self.input_weights = np.array(np.random.normal(0, 1, (8,)) * 0.1, dtype=np.float64)
        self.hidden_weights = np.array(np.random.normal(0, 1, (4,)) * 0.1, dtype=np.float64)
        self.input_bias = np.float64(np.random.normal(0, 1) * 0.1)
        self.hidden_bias = np.float64(np.random.normal(0, 1) * 0.1)
        
        self.training_time = 0
        self.accuracy = 0

    def sigmoid(self, x):
        # 명시적으로 float64 사용
        return np.float64(1.0) / (np.float64(1.0) + np.exp(-x, dtype=np.float64))

    def sigmoid_derivative(self, x):
        sig = self.sigmoid(x)
        return sig * (np.float64(1.0) - sig)

    def forward(self, x, y):
        # 입력을 float64로 변환
        x = np.float64(x)
        y = np.float64(y)
        
        # 입력층 -> 은닉층
        hidden = np.zeros(self.hidden_size, dtype=np.float64)
        for i in range(self.hidden_size):
            hidden[i] = self.sigmoid(x * self.input_weights[i] + 
                                   y * self.input_weights[i + self.hidden_size] + 
                                   self.input_bias)
        
        # 은닉층 -> 출력층
        output = np.float64(0)
        for i in range(self.hidden_size):
            output += hidden[i] * self.hidden_weights[i]
        output = self.sigmoid(output + self.hidden_bias)
        
        return output

    def train(self, data, labels, epochs=1000, learning_rate=0.1):
        # learning_rate를 float64로 변환
        learning_rate = np.float64(learning_rate)
        start_time = time.time()
        
        for epoch in range(epochs):
            total_loss = np.float64(0)
            
            for i in range(len(data)):
                x = np.float64(data[i][0])
                y = np.float64(data[i][1])
                target = np.float64(labels[i])
                
                # Forward pass
                hidden = np.zeros(self.hidden_size, dtype=np.float64)
                hidden_raw = np.zeros(self.hidden_size, dtype=np.float64)
                
                for j in range(self.hidden_size):
                    hidden_raw[j] = (x * self.input_weights[j] + 
                                   y * self.input_weights[j + self.hidden_size] + 
                                   self.input_bias)
                    hidden[j] = self.sigmoid(hidden_raw[j])
                
                output_raw = np.float64(0)
                for j in range(self.hidden_size):
                    output_raw += hidden[j] * self.hidden_weights[j]
                output_raw += self.hidden_bias
                output = self.sigmoid(output_raw)
                
                # 손실 계산
                loss = np.float64(0.5) * (target - output) * (target - output)
                total_loss += loss
                
                # Backward pass
                d_output = (output - target) * self.sigmoid_derivative(output_raw)
                
                # 은닉층 -> 출력층 가중치 업데이트
                for j in range(self.hidden_size):
                    self.hidden_weights[j] -= learning_rate * d_output * hidden[j]
                self.hidden_bias -= learning_rate * d_output
                
                # 입력층 -> 은닉층 가중치 업데이트
                for j in range(self.hidden_size):
                    d_hidden = (d_output * self.hidden_weights[j] * 
                              self.sigmoid_derivative(hidden_raw[j]))
                    self.input_weights[j] -= learning_rate * d_hidden * x
                    self.input_weights[j + self.hidden_size] -= learning_rate * d_hidden * y
                self.input_bias -= learning_rate * d_output
            
            if epoch % 100 == 0:
                print(f"Epoch {epoch}, Loss: {total_loss/len(data):.8f}")
        
        self.training_time = int((time.time() - start_time) * 1000)  # ms로 변환
        self.accuracy = self.get_accuracy(data, labels)

    def get_accuracy(self, data, labels):
        correct = 0
        for i in range(len(data)):
            output = self.forward(data[i][0], data[i][1])
            predicted = 1 if output >= 0.5 else 0
            if predicted == labels[i]:
                correct += 1
        return correct / len(data)

    def get_training_time(self):
        return self.training_time

def save_visualization(mlp, data, labels, filename):
    width = 300
    height = 250
    step = 5
    
    # 이미지 생성
    image = Image.new('RGB', (width, height), 'white')
    draw = ImageDraw.Draw(image)
    
    # 결정 경계 시각화 (배경 먼저)
    for x in range(0, width, step):
        for y in range(0, height, step):
            input_x = x / width
            input_y = y / height
            output = mlp.forward(input_x, input_y)
            
            # C++와 동일한 색상 계산
            r = int((1 - output) * 135 + output * 255)
            g = int((1 - output) * 206 + output * 105)
            b = int((1 - output) * 235 + output * 180)
            
            # 픽셀 채우기
            for dx in range(step):
                for dy in range(step):
                    if x + dx < width and y + dy < height:
                        draw.point((x + dx, y + dy), fill=(r, g, b))
    
    # 학습 데이터 포인트 그리기
    radius = 6
    for i in range(len(data)):
        x = int(data[i][0] * width)
        y = int(data[i][1] * height)
        color = (255, 0, 0) if labels[i] == 1 else (0, 0, 255)
        
        # 원 그리기
        draw.ellipse([x - radius, y - radius, x + radius, y + radius], fill=color)
    
    # 이미지 저장
    image.save(filename)

def main():
    start_time = time.time()
    
    print(f"Size of np.float64: {np.dtype(np.float64).itemsize} bytes")
    
    test_value = np.float64(0.1)
    bytes_array = test_value.tobytes()
    print("Internal representation of 0.1:", " ".join(f"{b:02x}" for b in bytes_array))
    print("--------------------------------")

    # 학습 데이터 생성 (C++와 동일)
    data = [
        [10, 23], [4, 21], [8, 17], [17, 22], [3, 12], [9, 12], [16, 14], [21, 20], [26, 22],  # upper
        [4, 4], [13, 7], [16, 6], [14, 3], [22, 8], [24, 5], [28, 6], [27, 14]                # lower
    ]
    
    labels = [1] * 9 + [0] * 8  # upper: 1, lower: 0
    
    # 데이터 정규화
    data = np.array(data, dtype=float)
    data[:, 0] /= 30.0  # x축: 0~30 기준
    data[:, 1] /= 25.0  # y축: 0~25 기준
    
    # MLP 모델 생성 및 학습
    mlp = MLP()
    mlp.train(data, labels)
    
    # 결과 시각화
    save_visualization(mlp, data, labels, "visualized_python.png")
    
    # 결과 출력
    total_time = int((time.time() - start_time) * 1000)
    train_time = mlp.get_training_time()
    print("--------------------------------")
    print("Saved Result: visualized_python.png")
    print(f"Train Accuracy: {int(100 * mlp.get_accuracy(data, labels))}%")
    print(f"Train Time: {train_time}ms")
    print(f"Overhead: {total_time - train_time}ms")
    print(f"Total Time: {total_time}ms")
    print()

if __name__ == "__main__":
    main()

File: mlp.cpp
#include "mlp.h"
#include <random>
#include <cmath>

double MLP::sigmoid(double x) const {
    return 1.0 / (1.0 + std::exp(-x));
}

double MLP::sigmoid_derivative(double x) const {
    double sig = sigmoid(x);
    return sig * (1.0 - sig);
}

double MLP::tanh(double x) const {
    return std::tanh(x);
}

double MLP::tanh_derivative(double x) const {
    double t = std::tanh(x);
    return 1.0 - t * t;
}

// Matrix multiplication: A @ x where A is matrix and x is vector
std::vector<double> MLP::matmul(const std::vector<std::vector<double>>& A,
                               const std::vector<double>& x) const {
    std::vector<double> result(A.size(), 0.0);
    for (size_t i = 0; i < A.size(); i++) {
        for (size_t j = 0; j < x.size(); j++) {
            result[i] += A[i][j] * x[j];
        }
    }
    return result;
}

// Outer product: a[:, None] @ b[None, :] in numpy
std::vector<std::vector<double>> MLP::outer(const std::vector<double>& a,
                                          const std::vector<double>& b) const {
    std::vector<std::vector<double>> result(a.size(), std::vector<double>(b.size()));
    for (size_t i = 0; i < a.size(); i++) {
        for (size_t j = 0; j < b.size(); j++) {
            result[i][j] = a[i] * b[j];
        }
    }
    return result;
}

MLP::MLP() 
    : hidden_size1(64),
      hidden_size2(64),
      input_weights(hidden_size1, std::vector<double>(2)),
      hidden_weights(hidden_size2, std::vector<double>(hidden_size1)),
      output_weights(hidden_size2),
      input_biases(hidden_size1),
      hidden_biases(hidden_size2),
      output_bias(0.0) {
    
    std::mt19937 gen(5489);  // Python과 동일한 시드값 사용
    
    double scale_input = std::sqrt(2.0 / 2);
    double scale_h = std::sqrt(2.0 / hidden_size1);
    double scale_out = std::sqrt(2.0 / hidden_size2);
    
    // Initialize input_weights (hidden_size1, 2)
    std::normal_distribution<> d_input(0, scale_input);
    for (auto& row : input_weights) {
        for (auto& w : row) {
            w = d_input(gen);
        }
    }
    
    // Initialize hidden_weights (hidden_size2, hidden_size1)
    std::normal_distribution<> d_h(0, scale_h);
    for (auto& row : hidden_weights) {
        for (auto& w : row) {
            w = d_h(gen);
        }
    }
    
    // Initialize output_weights (hidden_size2,)
    std::normal_distribution<> d_out(0, scale_out);
    for (auto& w : output_weights) {
        w = d_out(gen);
    }
    
    // Initialize biases to zeros
    std::fill(input_biases.begin(), input_biases.end(), 0.0);
    std::fill(hidden_biases.begin(), hidden_biases.end(), 0.0);
}

double MLP::forward(double x, double y) const {
    std::vector<double> input_vec = {x, y};
    
    // First hidden layer
    auto hidden1_raw = matmul(input_weights, input_vec);
    for (size_t i = 0; i < hidden1_raw.size(); i++) {
        hidden1_raw[i] += input_biases[i];
    }
    std::vector<double> hidden1(hidden1_raw.size());
    for (size_t i = 0; i < hidden1.size(); i++) {
        hidden1[i] = tanh(hidden1_raw[i]);
    }
    
    // Second hidden layer
    auto hidden2_raw = matmul(hidden_weights, hidden1);
    for (size_t i = 0; i < hidden2_raw.size(); i++) {
        hidden2_raw[i] += hidden_biases[i];
    }
    std::vector<double> hidden2(hidden2_raw.size());
    for (size_t i = 0; i < hidden2.size(); i++) {
        hidden2[i] = tanh(hidden2_raw[i]);
    }
    
    // Output layer
    double output_raw = 0.0;
    for (size_t i = 0; i < hidden_size2; i++) {
        output_raw += output_weights[i] * hidden2[i];
    }
    output_raw += output_bias;
    
    return sigmoid(output_raw);
}

void MLP::train(const std::vector<std::vector<double>>& data,
                const std::vector<double>& labels,
                int epochs,
                double learning_rate) {
    auto start = std::chrono::high_resolution_clock::now();
    
    for (int epoch = 0; epoch < epochs; epoch++) {
        double total_loss = 0.0;
        
        for (size_t idx = 0; idx < data.size(); idx++) {
            std::vector<double> input_vec = {data[idx][0], data[idx][1]};
            double target = labels[idx];
            
            // Forward pass
            auto hidden1_raw = matmul(input_weights, input_vec);
            for (size_t i = 0; i < hidden1_raw.size(); i++) {
                hidden1_raw[i] += input_biases[i];
            }
            std::vector<double> hidden1(hidden1_raw.size());
            for (size_t i = 0; i < hidden1.size(); i++) {
                hidden1[i] = tanh(hidden1_raw[i]);
            }
            
            auto hidden2_raw = matmul(hidden_weights, hidden1);
            for (size_t i = 0; i < hidden2_raw.size(); i++) {
                hidden2_raw[i] += hidden_biases[i];
            }
            std::vector<double> hidden2(hidden2_raw.size());
            for (size_t i = 0; i < hidden2.size(); i++) {
                hidden2[i] = tanh(hidden2_raw[i]);
            }
            
            double output_raw = 0.0;
            for (size_t i = 0; i < hidden_size2; i++) {
                output_raw += output_weights[i] * hidden2[i];
            }
            output_raw += output_bias;
            double output = sigmoid(output_raw);
            
            // Loss calculation
            double loss = 0.5 * (target - output) * (target - output);
            total_loss += loss;
            
            // Backward pass
            double d_output = (output - target) * sigmoid_derivative(output_raw);
            
            // d_hidden2: element-wise multiplication
            std::vector<double> d_hidden2(hidden_size2);
            for (size_t i = 0; i < hidden_size2; i++) {
                d_hidden2[i] = d_output * output_weights[i] * tanh_derivative(hidden2_raw[i]);
            }
            
            // d_hidden1: matrix multiplication and element-wise multiplication
            std::vector<double> d_hidden1(hidden_size1, 0.0);
            for (size_t i = 0; i < hidden_size1; i++) {
                for (size_t j = 0; j < hidden_size2; j++) {
                    d_hidden1[i] += d_hidden2[j] * hidden_weights[j][i];
                }
                d_hidden1[i] *= tanh_derivative(hidden1_raw[i]);
            }
            
            // Update weights and biases
            // Update output layer
            for (size_t i = 0; i < hidden_size2; i++) {
                output_weights[i] -= learning_rate * d_output * hidden2[i];
            }
            output_bias -= learning_rate * d_output;
            
            // Update hidden layer
            auto d_hidden2_outer = outer(d_hidden2, hidden1);
            for (size_t i = 0; i < hidden_size2; i++) {
                for (size_t j = 0; j < hidden_size1; j++) {
                    hidden_weights[i][j] -= learning_rate * d_hidden2_outer[i][j];
                }
                hidden_biases[i] -= learning_rate * d_hidden2[i];
            }
            
            // Update input layer
            auto d_hidden1_outer = outer(d_hidden1, input_vec);
            for (size_t i = 0; i < hidden_size1; i++) {
                for (size_t j = 0; j < 2; j++) {
                    input_weights[i][j] -= learning_rate * d_hidden1_outer[i][j];
                }
                input_biases[i] -= learning_rate * d_hidden1[i];
            }
        }
        
        if (epoch % 100 == 0) {
            std::cout << "Epoch " << epoch << ", Loss: " << total_loss/data.size() << std::endl;
        }
    }
    
    auto end = std::chrono::high_resolution_clock::now();
    training_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
}

double MLP::getAccuracy(const std::vector<std::vector<double>>& data,
                       const std::vector<double>& labels) const {
    int correct = 0;
    for (size_t i = 0; i < data.size(); i++) {
        double output = forward(data[i][0], data[i][1]);
        if ((output >= 0.5 && labels[i] == 1) || (output < 0.5 && labels[i] == 0)) {
            correct++;
        }
    }
    return static_cast<double>(correct) / data.size();
}

double MLP::getTrainingTime() const {
    return training_time;
}
